stl/inc/format

```diff
@@ -2269,1020 +2335,1179 @@ using _Fmt_wit = back_insert_iterator<_Fmt_buffer<wchar_t>>;
+            _Measure_string_prefix_iterator_legacy, _Measure_string_prefix_iterator_utf<_CharT>>;
 
-    constexpr _Measure_string_prefix_iterator_utf() = default;
+    template <class _CharT>
+    _NODISCARD const _CharT* _Measure_string_prefix(const basic_string_view<_CharT> _Value, int& _Width) {
+        // Returns a pointer past-the-end of the largest prefix of _Value that fits in _Width, or all
+        // of _Value if _Width is negative. Updates _Width to the estimated width of that prefix.
+        const int _Max_width = _Width;
+        const auto _First    = _Value.data();
+        const auto _Last     = _First + _Value.size();
+        _Measure_string_prefix_iterator<_CharT> _Pfx_iter(_First, _Last);
+        int _Estimated_width = 0; // the estimated width of [_First, _Pfx_iter)
+
+        constexpr auto _Max_int = (numeric_limits<int>::max)();
+
+        while (_Pfx_iter != default_sentinel) {
+            if (_Estimated_width == _Max_width && _Max_width >= 0) {
+                // We're at our maximum length
+                break;
+            }
 
-    _NODISCARD constexpr value_type operator*() const {
-        return _Unicode_width_estimate(*_WrappedIter);
-    }
+            const int _Character_width = *_Pfx_iter;
 
-    constexpr _Measure_string_prefix_iterator_utf& operator++() noexcept {
-        ++_WrappedIter;
-        return *this;
-    }
+            if (_Max_int - _Character_width < _Estimated_width) { // avoid overflow
+                // Either _Max_width isn't set, or adding this character will exceed it.
+                if (_Max_width < 0) { // unset; saturate width estimate and take all characters
+                    _Width = _Max_int;
+                    return _Last;
+                }
+                break;
+            }
 
-    constexpr _Measure_string_prefix_iterator_utf operator++(int) noexcept {
-        auto _Old = *this;
-        ++*this;
-        return _Old;
-    }
+            _Estimated_width += _Character_width;
+            if (_Estimated_width > _Max_width && _Max_width >= 0) {
+                // with this character, we exceed the maximum length
+                _Estimated_width -= _Character_width;
+                break;
+            }
+            ++_Pfx_iter;
+        }
 
-    _NODISCARD constexpr const _CharT* _Position() const {
-        return _WrappedIter._Position();
+        _Width = _Estimated_width;
+        return _Pfx_iter._Position();
     }
-};
 
-class _Measure_string_prefix_iterator_legacy {
-private:
-    _Fmt_codec<char> _Codec = _Get_fmt_codec<char>();
-    const char* _First      = nullptr;
-    const char* _Last       = nullptr;
-    int _Units              = 0;
-
-    void _Update_units() {
-        if (_First < _Last) {
-            _Units = _Codec._Units_in_next_character(_First, _Last);
+    template <class _CharT>
+    _NODISCARD constexpr const _CharT* _Choose_literal(const char* const _Str, const wchar_t* const _WStr) noexcept {

```

Comment:
It appears that `_Choose_literal` and `_STATICALLY_WIDEN` should appear within the `#if _HAS_CXX23` block, instead of outside.

And actually, this duplicates `<chrono>`:

https://github.com/microsoft/STL/blob/5404ba9c25f26f25a0ac50e6c4defce7833a8da6/stl/inc/chrono#L4692-L4704

It's more concerning because `<format>` doesn't `#undef` (while `<chrono>` does).

Can we unify these? We need it for C++23 and `<chrono>` needs it for concepts, so we could just say "C++23 or concepts", and find a central location.

Reply:
I feel that `<format>` is the best location (unless we introduce another internal header). `<chrono>` includes `<format>` when `__cpp_lib_concepts` is defined, so both `<format>` and `<chrono>` can use `_STATICALLY_WIDEN` when it's defined here.

I moved `_STATICALLY_WIDEN` to the start of `<format>` and added a comment to note that it's also used in `<chrono>`.

Reply:
I see, I was confused last night - you were already unifying them! The new location and comment are much appreciated.